File: z3ext.formatter/trunk/src/z3ext/formatter/humandatetime.py
##############################################################################
#
# Copyright (c) 2009 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" ``humanDatetime`` formatter implementation

$Id$
"""
from pytz import utc, timezone
from datetime import datetime
from zope import interface, component
from zope.i18n import translate
from zope.component import getUtility
from zope.publisher.interfaces.http import IHTTPRequest

from interfaces import IFormatter, IFormatterFactory, IFormatterConfiglet


class HumanDatetimeFormatter(object):
    interface.implements(IFormatter)

    def __init__(self, request, *args):
        self.request = request

    def format(self, value):
        configlet = getUtility(IFormatterConfiglet)
        tz = timezone(configlet.timezone)

        if value.tzinfo is None:
            value = datetime(value.year, value.month, value.day, value.hour,
                             value.minute, value.second, value.microsecond, tz)

        value = value.astimezone(tz)
        delta = datetime.now(utc) - value.astimezone(utc)

        years, months, weeks, hours, minutes = (
            delta.days/365, delta.days/30, delta.days/7,
            delta.seconds/3600, delta.seconds/60)
        formatted = None
        if years > 0:
            formatted = translate(
                u'${value} year(s) ago', 'z3ext.formatter',
                mapping={'value': years})

        if months > 0:
            formatted = translate(u'${value} month(s) ago', 'z3ext.formatter',
                             mapping={'value': months})
        elif weeks > 0:
            formatted = translate(u'${value} week(s) ago', 'z3ext.formatter',
                             mapping={'value': weeks})
        elif delta.days > 0:
            formatted = translate(u'${value} day(s) ago', 'z3ext.formatter',
                             mapping={'value': delta.days})
        elif hours > 0:
            formatted = translate(u'${value} hour(s) ago', 'z3ext.formatter',
                             mapping={'value': hours})
        elif minutes > 0:
            formatted = translate(u'${value} minute(s) ago', 'z3ext.formatter',
                             mapping={'value': minutes})
        else:
            formatted = translate(u'${value} second(s) ago', 'z3ext.formatter',
                         mapping={'value': delta.seconds})

        return """<span class="z3ext-formatter-humandatetime" value="%s">%s</span>""" \
                % (value.strftime('%Y %B %d %H:%M:%S %Z'), formatted)


class HumanDatetimeFormatterFactory(object):
    component.adapts(IHTTPRequest)
    interface.implements(IFormatterFactory)

    def __init__(self, request):
        self.request = request

    def __call__(self, *args, **kw):
        return HumanDatetimeFormatter(self.request, *args)

File: z3ext.formatter/trunk/src/z3ext/formatter/resources/humandatetime.js
var my_dictionary = {
    'year(s) ago'  : 'year(s) ago',
    'month(s) ago'  : 'month(s) ago',
    'weeks(s) ago': 'weeks(s) ago',
    'days(s) ago' : 'days(s) ago',
    'hours(s) ago' : 'hours(s) ago',
    'minutes(s) ago' : 'minutes(s) ago',
    'second(s) ago' : 'second(s) ago'
};

$.i18n.setDictionary(my_dictionary);

function setFormatter(el)
{
    var time = new Date()
    time.setTime(Date.parse(el.attr('value')));
    var new_time = new Date();
    delta = new_time-time;
    years = Math.floor(delta/(365*24*60*60*1000.0));
    months = Math.floor(delta/(30*24*60*60*1000.0));
    weeks = Math.floor(delta/(7*24*60*60*1000.0));
    days = Math.floor(delta/(24*60*60*1000.0));
    hours = Math.floor(delta/(60*60*1000.0));
    minutes = Math.floor(delta/(60*1000.0));
    seconds = Math.floor(delta/(1000.0));
    var res;
    if (years)
        res = years + ' ' +$.i18n._('year(s) ago')
    else if (months)
        res = months + ' ' +$.i18n._('month(s) ago')
    else if (weeks)
        res = weeks + ' ' +$.i18n._('weeks(s) ago')
    else if (days)
        res = days + ' ' +$.i18n._('days(s) ago')
    else if (hours)
        res = hours + ' ' +$.i18n._('hours(s) ago')
    else if (minutes)
        res = minutes + ' ' +$.i18n._('minutes(s) ago')
    else
        res = seconds + ' ' +$.i18n._('second(s) ago')
    el.text(res);
    el.attr('processed', 'true')
}


$(document).ready(function() {
    var elems = $("span.z3ext-formatter-humandatetime");
    for (var i = 0; i < elems.length; i++)
    {
        var el = $(elems[i]);
        if (!el.attr('processed')) {
            setFormatter(el);
        }
    }
});

/*
d1 = datetime.now(utc)
d2 = value.astimezone(utc)

delta = d1 - d2

years, months, weeks, hours, minutes = (
    delta.days/365, delta.days/30, delta.days/7,
    delta.seconds/3600, delta.seconds/60)

if years > 0:
    return translate(
        u'${value} year(s) ago', 'z3ext.formatter',
        mapping={'value': years})

if months > 0:
    return translate(u'${value} month(s) ago', 'z3ext.formatter',
                     mapping={'value': months})

if weeks > 0:
    return translate(u'${value} week(s) ago', 'z3ext.formatter',
                     mapping={'value': weeks})

if delta.days > 0:
    return translate(u'${value} day(s) ago', 'z3ext.formatter',
                     mapping={'value': delta.days})

if hours > 0:
    return translate(u'${value} hour(s) ago', 'z3ext.formatter',
                     mapping={'value': hours})

if minutes > 0:
    return translate(u'${value} minute(s) ago', 'z3ext.formatter',
                     mapping={'value': minutes})

return translate(u'${value} second(s) ago', 'z3ext.formatter',
                 mapping={'value': delta.seconds})
*/